Assets/Scripts/Models/State/JobState.cs

```diff
@@ -0,0 +1,328 @@
+                }
+            }
+            else
+            {
+                path = Pathfinding.FindPathToFurniture(character.CurrTile, Job.JobObjectType);    
+            }
+
+            if (path == null || path.Count == 0)
+            {
+                // We can't find a furniture of the right type so abandon job, exit this state and stop processing
+                AbandonJob();
+                Finished();
+                return true;
+            }
+
+            Job.tile = path.Last();
+            if (Job.IsTileAtJobSite(character.CurrTile))
+            {
+                // We are already there!
+                return false;
+            }
+
+            FSMLog(" - Need to walk to furniture");
+
+            // Let's go there
+            MoveState moveState = new MoveState(character, Job.IsTileAtJobSite, path, this);
+            character.SetState(moveState);
+            return true;
+        }
+
+        // Returns true if everything has been handled
+        private bool NeedToFindMoreMaterial()
+        {
+            // Checks if material is already delivered to the job site
+            if (Job.MaterialNeedsMet())
+            {
+                return false;
+            }
+
+            // Check if character is carrying something that is needed
+            if (character.inventory != null)
+            {
+                int amountNeeded = Job.AmountDesiredOfInventoryType(character.inventory);
+
+                // 0 means that they don't want that inventory and we need to dump!
+                if (amountNeeded == 0)
+                {
+                    DumpExcessInventory();
+
+                    // Restart method as a stop gap measure instead of hauling
+                    // TODO: Change this when DumpExcessInventory() creates a hauling job
+                    return NeedToFindMoreMaterial();
+                }
+
+                // We don't have enough, but we can carry enough
+                if (amountNeeded > character.inventory.StackSize && character.inventory.MaxStackSize > amountNeeded)
+                {
+                    return FetchInventory(character.inventory.ObjectType, amountNeeded - character.inventory.StackSize);
+                }
+
+                // We don't have enough, but we need to deliver what we have
+                return false;
+            }
+
+            // At this point we know we don't carry anything, so we can grab anything required
+            // TODO: GetFirstFulfillableInventoryRequirement doesn't do the stockpile and lock check.
+            Inventory neededInventory = Job.GetFirstFulfillableInventoryRequirement();
+            if (neededInventory == null)
+            {
+                AbandonJob();
+                Finished();
+                return true;
+            }
+
+            return FetchInventory(neededInventory.ObjectType, neededInventory.StackSize);
```

Comment:
so this should be probably:
`return FetchInventory(neededInventory.ObjectType, neededInventory.MaxStackSize - neededInventory.StackSize);`
code when character has something in Inv. is handled above that